### PR TITLE
[WIP] ✨ Add support for Ignition v3.x

### LINKED ISSUE #9157: Ignition v3.x support
### What would you like to be added (User Story)?

As a user I would like to bootstrap Flatcar & FCOS with Ignition v3.x as Ignition v2.x doesn't work with FCOS and is only required for flatcar LTS release. 

### Detailed Description

The current implementation (Container Linux Config aka clc) is used when setting `format` to "ignition" in `KubeadmConfigSpec`.
To be able to support both v2.x  & v3.x, a new field needs to be added to know which transpiler should be used:
- clc for v2.x
- butane for v3.x

Like in the current implementation, an additional butane config may be provided to extend/override the generated Ignition config.

### Anything else you would like to add?

This issue follows this slack conversation: https://kubernetes.slack.com/archives/C8TSNPY4T/p1688643597808979

### Label(s) to be applied

/kind feature
/area bootstrap
/area provider/bootstrap-kubeadm

### PR DESCRIPTION
**Pull request by bengentil**
**What this PR does / why we need it**:

Implements Ignition v3.x

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #9157

### CODE REVIEW

**Comment by bengentil** [comment]
Hi @invidian,

To follow the slack conversation we had, I pushed this PR so we can discuss the implementation in details.

The PR is working but I have a few questions:

1. I've added a new optional enum `Transpiler` and a `ButaneConfig` (similar to `ContainerLinuxConfig`), is this how you pictured the switch between clc & butane?
2. The butane implementation is very similar to the clc implementation, I think some templating functions can be shared (`defaultTemplateFuncMap`, `mountpointName`, `templateYAMLIndent`, `parseOwner`) in maybe a third package, for the rest I think it's way simpler and readable to keep 2 separate implementation, what do you think?

**Review by invidian — CHANGES_REQUESTED** [review]
Hey @bengentil, thanks for opening this PR, this is highly appreciated! To answer your questions:
> I've added a new optional enum Transpiler and a ButaneConfig (similar to ContainerLinuxConfig), is this how you pictured the switch between clc & butane?

Not really, I explained that in more details in a comment.

>The butane implementation is very similar to the clc implementation, I think some templating functions can be shared (defaultTemplateFuncMap, mountpointName, templateYAMLIndent, parseOwner) in maybe a third package, for the rest I think it's way simpler and readable to keep 2 separate implementation, what do you think?

There is naturally a lot in common indeed, so definitely putting common parts in one place would be great, although I do not treat it as a must have.

However, looking at https://github.com/kubernetes-sigs/cluster-api/blob/aac6da753c002e04e261823d4365fe7bcccf95c7/bootstrap/kubeadm/api/v1beta1/kubeadmconfig_webhook.go#L210, I'm wondering if perhaps Butane does not allow us to translate more features of cloudconfig into Ignition, which would introduce some differences between Butane and CLC. This is perhaps something we could look at.

Overall PR looks in a good shape to me, it's great to see it finally happening. Once it lands, perhaps Flatcar templates can be updated to use it so we can gradually phase out CLC and also Infra providers will be able to add support for FCOS.

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go
Comment (bengentil) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:123]: I think this is required for OS who doesn't support both implementation (v2.x only like Flatcar LTS, v3.x only like FCOS)

Relying on `ContainerLinuxConfig` and `ButaneConfig` being not nil is not sufficient as both can be nil (eg. I want to install FCOS with no additional config)

Comment (invidian) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:123]: We can also make one of them required in Ignition type then, so user have to choose? We might also want to document that Butane supports generating Ignition v3 while CLC only v2.

IMO it boils down to: `clc: {}`/`butane: {}` vs `transpiler: clc`/`transpiler: bunane`, but for me the latter makes API more complex as there are 2 knobs for controlling one thing and it creates a matrix of additional bad configurations e.g. `transpiler: clc` with `butane: {}` in addition to existing scenarios like `butane: {}` and `clc: {}`, which we cannot avoid (they are essential).

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go
Comment (bengentil) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go]: Good catch! fixed

File: bootstrap/kubeadm/internal/ignition/butane/butane.go
Comment (bengentil) [reply bootstrap/kubeadm/internal/ignition/butane/butane.go:56]: This is the variant & version of the static template only, the additionalConfig must specify it's own variant/version, both are translated to ignition v3.4.0 and the merge is done in ignition v3.4.0.

File: bootstrap/kubeadm/internal/ignition/butane/butane.go
Comment (bengentil) [reply bootstrap/kubeadm/internal/ignition/butane/butane.go:57]: Same as above

File: bootstrap/kubeadm/internal/ignition/butane/butane.go
Comment (invidian) [reply bootstrap/kubeadm/internal/ignition/butane/butane.go:56]: That seem counter-intuitive, as `additionalConfig` name suggests it's _optional_ and so is the API semantics for this field. How do we _require_ it right now? Shouldn't we require this information from user and use it then for both base template and additional config?

File: bootstrap/kubeadm/internal/ignition/butane/butane.go
Comment (bengentil) [reply bootstrap/kubeadm/internal/ignition/butane/butane.go:52]: Indeed, fixed

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go
Comment (bengentil) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:123]: I understand you point but what I propose is basically the same as we have in `KubeadmConfigSpec`:
- a field `Format` to switch between "cloud-init" and "ignition"
- a field `Ignition` used to pass optional ignition configs only when `Format` is explicitly set to "ignition"

Knowing that the `Ignition` field is optional, making one of `clcConfig`/`butaneConfig` required would means any existing declaration without additional config would have to add:
```
ignition:
   clcConfig: {}
```
(or `clc` if the field is renamed)
And when ignition v2.x is phased out from flatcar, a new change would be required to:
```
ignition:
  butaneConfig: {}
```

I'm not sure this is a good UX, what I proposed initially was an enum `version` field to be self-explanatory about the Ignition version the user want (and let the implementation decide the correct transpiler to call), I changed it to a transpiler field to be consistent with the `clcConfig` and `butaneConfig` fields but the long-term logic is the same:
- first we introduce a new optional transpiler, so any user who want to start using Ignition v3 can (no impact on existing manifests)
- once Flatcar LTS updates to Ignition v3, butane becomes the default transpiler (when declared, additionalConfig must be moved from clcConfig to butaneConfig in a variant/version supported by butane or the `transpiler` field should be explicitly set to clc)
- at some point, clc implementation is phased out

File: bootstrap/kubeadm/internal/ignition/butane/butane.go
Comment (bengentil) [reply bootstrap/kubeadm/internal/ignition/butane/butane.go:57]: `additionalConfig` is still optional, the variant/version is required by butane when `additionalConfig` is not nil to transpile it to Ignition v3.4.0.

This version doesn't have to be the same as the one used in the static template as the merge is done in Ignition v3.4.0, so a user can add an  `additionalConfig`  in any variant/version they variant (as long as the couple can be transpiled to Ignition v3.4.0)

We could add a default variant/version if none is specified as it would work in most cases.

File: bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go
Comment (invidian) [reply bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:123]: That's a good point, I didn't think about the backward compatibility initially. So how about still picking CLC as a default if both `clcConfig` and `butaneConfig` are `nil`? This remains backward compatible and users which needs Ignition v3 can specify `butaneConfig: {}`.

It does however make selection of Ignition version kind of ambiguous, which I'm not a fan of. Having an enum for Ignition version with `v2` default for the time being and `v3` as a new option as you originally suggested does make sense in this case. So for users without additional configuration, we would use CLC as a default transpiler for v2 and Butane for v3, but that would be an implementation detail. I also wonder if it make sense to support full versions like `2.3.0` or we could get away with just v2/v3 distinction for the time being.

At the same time, requesting Ignition v3 with CLC configuration would result in an error, as CLC is not capable of producing it. The similar validation would need to be added for Butane. 

It would be good for someone else to tip in, maybe @pothos or @jepio could have a look with their critical eye :) And definitely someone from CAPI maintainers.

File: bootstrap/kubeadm/internal/ignition/butane/butane.go
Comment (invidian) [reply bootstrap/kubeadm/internal/ignition/butane/butane.go:57]: I understand that, but if user specifies no `additionalConfig`, why should they be forced to use `flatcar` variant in this case? What if this variant is not compatible with their OS?

**Comment by fabriziopandini** [comment]
If I remember well when we implemented ingition we agreed that this was a temporary solution, while a more long-term solution for machine/node bootstrap should be worked out in order to avoid CABPK becoming unmaintainable, its API exploding, the main CAPI repo to pull too many dependency.

I think we should revive this discussion before moving forward with this PR, before adding complexity on top of a situation that we know already not ideal

cc @vincepri @CecileRobertMichon @enxebre @sbueringer @killianmuldoon @invidian @dongsupark @johananl

**Comment by johananl** [comment]
Yes @fabriziopandini, I think this is the relevant issue: https://github.com/kubernetes-sigs/cluster-api/issues/5294

**Comment by johananl** [comment]
IMO we can freeze/close this PR for now. #5294 is WIP and AFAICT we can't advance this PR until that design proposal is settled.

**Comment by vincepri** [comment]
@fabriziopandini @johananl @invidian @bengentil Ignition v2 has long standing issues as today, given that these APIs are still behind a feature gate, I can see why adding support for v3 going forward would be beneficial.

That said, I'd personally like to see a more cohesive approach to Ignition support within Cluster API, on the Machine object specifically.

**Comment by johananl** [comment]
>That said, I'd personally like to see a more cohesive approach to Ignition support within Cluster API, on the Machine object specifically.

Yes, this is one of the core areas we're trying to address in #5294. I'm definitely taking that into account in my WIP design proposal. I'll share a draft as soon as I can.

**Comment by sbueringer** [comment]
What is the current status of this PR? I'm not sure what the next steps are to make progress with this PR/topic.

**Comment by fabriziopandini** [comment]
Based on the latest discussion in the community I'm aware of about this topic, we should wait for progress on https://github.com/kubernetes-sigs/cluster-api/issues/5294 first, but feel free to bring it up again in the office hours given that some time is passed now and folks might have different opinions 

/close

**Comment by johananl** [comment]
Just for context, #5294 touches some very basic assumptions in the CAPI design and it will therefore likely take some time to agree on the design and finish the implementation.

**Pull request closed without merging**
